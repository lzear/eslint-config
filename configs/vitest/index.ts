/*
  eslint perfectionist/sort-objects: [
    'error',
    {
      type: 'alphabetical',
      order: 'asc',
      partitionByNewLine: true
    }
  ]
*/

import type { Linter } from 'eslint'

import type { ConfigOptions } from '../..'

import { interopDefault } from '../../utils'

export const vitest = async (config: ConfigOptions): Promise<Linter.Config> => {
  if (!config.vitest) {
    return {}
  }

  const vitestPlugin = await interopDefault(import('@vitest/eslint-plugin'))

  const files = [
    '**/test/*.js',
    '**/test/*.cjs',
    '**/test/*.mjs',
    '**/*.test.js',
    '**/*.test.cjs',
    '**/*.test.mjs',
  ]

  if (config.typescript) {
    files.push(
      '**/test/*.ts',
      '**/test/*.cts',
      '**/test/*.mts',
      '**/*.test.ts',
      '**/*.test.cts',
      '**/*.test.mts',
    )
  }

  if (config.react) {
    files.push('**/test/*.jsx', '**/*.test.jsx')

    if (config.typescript) {
      files.push('**/test/*.tsx', '**/*.test.tsx')
    }
  }

  return {
    name: 'lzear/vitest',

    files,

    plugins: {
      vitest: vitestPlugin,
    },

    rules: { ...vitestPlugin.configs.recommended.rules },

    settings: {
      vitest: {
        typecheck: true,
      },
    },
  }
}
