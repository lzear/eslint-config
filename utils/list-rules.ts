import fs from 'node:fs'
import path from 'node:path'

import configBase from '@lzear/eslint-config'
import configReact from '@lzear/eslint-config-react'
import configTs from '@lzear/eslint-config-typescript'
import { ESLint, Linter } from 'eslint'
import _ from 'lodash'

export const listRules = async (overrideConfig: Linter.Config) => {
  const eslint = new ESLint({ overrideConfig: overrideConfig })
  const filePath = './a/somefile.js'
  const config = await eslint.calculateConfigForFile(filePath)
  const warnings: string[] = []
  const errors: string[] = []
  for (const [rule, config1] of Object.entries(config.rules)) {
    if (Array.isArray(config1)) {
      switch (config1[0]) {
        case 0: {
          break
        }
        case 1: {
          warnings.push(rule)
          break
        }
        case 2: {
          errors.push(rule)
          break
        }
        default: {
          throw new Error(`Unexpected rule configuration (${rule})`)
        }
      }
    } else {
      throw new TypeError(`Rule configuration must be an object (${rule})`)
    }
  }
  return { warnings, errors }
}

export const rulesDir = path.resolve(import.meta.dirname, '../rules')

export const allRules = async () => {
  const rulesObjs = await Promise.all(
    [configBase, configTs, configReact].map((config) =>
      listRules(config as Linter.Config),
    ),
  )
  const rules = []
  for (const { warnings, errors } of rulesObjs)
    rules.push(...warnings, ...errors)
  return _.uniq(rules).sort()
}

export const testedRules = async () => {
  const getFolders = async (
    rootPath: string,
    folders: string[] = [],
  ): Promise<string[]> => {
    const files = await fs.promises.readdir(
      path.resolve(rootPath, ...folders),
      { withFileTypes: true },
    )
    const dirs = files.filter((file) => file.isDirectory())
    if (dirs.length === 0) return [path.join(...folders)]
    const r = await Promise.all(
      dirs.map((dir) => getFolders(rootPath, [...folders, dir.name])),
    )
    return r.flat()
  }
  return getFolders(rulesDir)
}

export const undocumentedRules = async () => {
  const documented = await testedRules()
  const all = await allRules()
  return _.difference(all, documented).sort()
}
