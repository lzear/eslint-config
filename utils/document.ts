import fs from 'node:fs'
import path from 'node:path'

import { builtinRules } from 'eslint/use-at-your-own-risk'

import { rulesDir } from './list-rules'

const ruleMetaDic = {
  unicorn: async (rule: string) => {
    const r = await import(`eslint-plugin-unicorn/rules/${rule}`)
    return r.default.meta
  },
  js: async (rule: string) => {
    const r = builtinRules.get(rule)
    if (!r) throw new Error(`Unknown rule: ${rule}`)
    return r.meta
  },
}

const getRuleMeta = async (rule: string) => {
  const [name, domain] = rule.split('/').reverse()
  if (!name) throw new Error(`Invalid rule: ${rule}`)
  // @ts-expect-error object might miss key
  const getMeta = ruleMetaDic[domain || 'js']
  if (!getMeta) throw new Error(`Unknown domain: ${domain}`)
  return getMeta(name)
}

export const documentRule = async (rule: string) => {
  const ruleDir = path.resolve(rulesDir, rule)

  const [good, bad] = await Promise.all([
    fs.promises.readFile(path.resolve(ruleDir, 'good.ts'), 'utf8'),
    fs.promises.readFile(path.resolve(ruleDir, 'bad.ts'), 'utf8'),
  ])

  return { good, bad, meta: await getRuleMeta(rule) }
}
