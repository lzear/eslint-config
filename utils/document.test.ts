import { describe, expect, it } from 'vitest'

import { documentRule } from './document'
import { testedRules } from './list-rules'

describe(documentRule, () => {
  it('documents one rule', async () => {
    expect(await documentRule('unicorn/better-regex')).toStrictEqual({
      good: 'const a = /\\d/\nconst b = /\\D/\nconst c = /\\w/\nconst d = /\\w/i\nconst e = /\\W/\nconst f = /\\W/i\nconst g = /\\d\\.\\w-\\D/i\n',
      bad: 'const a = /[0-9]/\nconst b = /[^0-9]/\nconst c = /[a-zA-Z0-9_]/\nconst d = /[a-z0-9_]/i\nconst e = /[^a-zA-Z0-9_]/\nconst f = /[^a-z0-9_]/i\nconst g = /[0-9]\\.[a-zA-Z0-9_]\\-[^0-9]/i\n',
      meta: {
        type: 'suggestion',
        docs: {
          description:
            'Improve regexes by making them shorter, consistent, and safer.',
        },
        fixable: 'code',
        schema: [
          {
            type: 'object',
            additionalProperties: false,
            properties: {
              sortCharacterClasses: {
                type: 'boolean',
                default: true,
              },
            },
          },
        ],
        messages: {
          'better-regex': '{{original}} can be optimized to {{optimized}}.',
          'better-regex/parse-error': 'Problem parsing {{original}}: {{error}}',
        },
      },
    })
  })

  it('documents all rules', async () => {
    const documented = await testedRules()
    expect(
      await Promise.all(documented.map((r) => documentRule(r))),
    ).toMatchSnapshot()
  })
})
