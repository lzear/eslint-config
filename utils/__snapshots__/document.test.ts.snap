// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`documentRule > documents all rules 1`] = `
[
  {
    "bad": "// @ts-nocheck

class A extends B {
  constructor() {} // Would throw a ReferenceError.
}

// Classes which inherits from a non constructor are always problems.
class C extends null {
  constructor() {
    super() // Would throw a TypeError.
  }
}

class D extends null {
  constructor() {} // Would throw a ReferenceError.
}
",
    "good": "// @ts-nocheck

class A {
  constructor() {}
}

class B extends C {
  constructor() {
    super()
  }
}
",
    "meta": {
      "docs": {
        "description": "Require \`super()\` calls in constructors",
        "recommended": true,
        "url": "https://eslint.org/docs/latest/rules/constructor-super",
      },
      "messages": {
        "badSuper": "Unexpected 'super()' because 'super' is not a constructor.",
        "duplicate": "Unexpected duplicate 'super()'.",
        "missingAll": "Expected to call 'super()'.",
        "missingSome": "Lacked a call of 'super()' in some code paths.",
      },
      "schema": [],
      "type": "problem",
    },
  },
  {
    "bad": "const a = /[0-9]/
const b = /[^0-9]/
const c = /[a-zA-Z0-9_]/
const d = /[a-z0-9_]/i
const e = /[^a-zA-Z0-9_]/
const f = /[^a-z0-9_]/i
const g = /[0-9]\\.[a-zA-Z0-9_]\\-[^0-9]/i
",
    "good": "const a = /\\d/
const b = /\\D/
const c = /\\w/
const d = /\\w/i
const e = /\\W/
const f = /\\W/i
const g = /\\d\\.\\w-\\D/i
",
    "meta": {
      "docs": {
        "description": "Improve regexes by making them shorter, consistent, and safer.",
      },
      "fixable": "code",
      "messages": {
        "better-regex": "{{original}} can be optimized to {{optimized}}.",
        "better-regex/parse-error": "Problem parsing {{original}}: {{error}}",
      },
      "schema": [
        {
          "additionalProperties": false,
          "properties": {
            "sortCharacterClasses": {
              "default": true,
              "type": "boolean",
            },
          },
          "type": "object",
        },
      ],
      "type": "suggestion",
    },
  },
]
`;
